Workstation wide cache and RRF

cstamas opened the pull request:
The build (especially with empty local repo) shows how Maven "wanders around" to multiple remote repositories to take stuff that is available from Central proper.

Also, some dependencies introduce new remote repositories making this even more painful (time consuming).

This PR adds two things:
* Mimir Cache: `~/.mimir/local` is a "pure" cache, unlike Maven Local Repo that is "mixed bag" of installed/cached artifacts.
* Maven RRF (3.9+ feature): limiting where Maven can go (currently all allowed groupIDs are empty, as Central has all)

These two now makes "-P fast" build below 20 seconds on my workstation **with empty local repository** (and Mimir caches primed).

Moreover, CI caching now becomes trivial, as between jobs it is only `~/.mimir/local` that needs to be saved and restored before next run. This way even Central is interacted minimally (only new deps are pulled, once).

cstamas committed:
Workstation wide cache and RRF

The build (especially with empty local repo) shows how
Maven "wanders around" to multiple remote repositories
to take stuff that is available from Central proper.

Also, some dependencies introduce new remote repositories
making this even more painful (time consuming).

This PR adds two things:
* Mimir Cache: (~/.mimir/local is a "pure" cache,
  unlike local repo that is "mixed bag" of installed/cached artifacts.
* Maven RRF (3.9+ feature): limiting where Maven can go
  (currently all allowed groupIDs are empty, as Central has all)

These two now makes "-P fast" build below 20 seconds on
my workstation _with empty local repository_ (and Mimir
caches primed).

Moreover, CI caching now becomes trivial, as between
jobs it is only ~/.mimir/local that needs to be saved
and restored before next run. This way even Central is
interacted minimally (only new deps are pulled, once).

cstamas commented:
All the files `groupId-XXX.txt` are **empty** as Maven should not even sniff them, as everything needed is on Central. This may need tuning, for example for SNAPSHOT artifacts (as I saw them in POM), but given there is no file for them, filtering is not applied either to them.

joakime commented:
> Moreover, CI caching now becomes trivial, as between jobs it is only `~/.mimir/local` that needs to be saved and restored before next run. This way even Central is interacted minimally (only new deps are pulled, once).

How would you change https://github.com/jetty/jetty.project/blob/jetty-12.0.x/.github/workflows/codeql-analysis.yml to take advantage of this cache?

cstamas commented:
This is how we do it with Maven:
https://github.com/apache/maven/blob/master/.github/workflows/maven.yml#L44-L58

Basically just restore the cache, and if extension is present, it will use it.

cstamas commented:
Let me try to modify this GH action to see. Will let current job finish (codeQL) as we will have baseline w/o caches.

cstamas commented:
Hm, in fact we have a fairly recent "baseline":
![image](https://github.com/user-attachments/assets/9aa100a0-e909-45fe-b0bd-8d91a236e5c0)

cstamas committed:
Use mimir caches in CodeQL analysis

cstamas commented:
This build will now populate Mimir caches, no time change expected...

cstamas commented:
Sadly we tossed previous (first in this PR) build cache, but at build end Mimir [reported this](https://github.com/jetty/jetty.project/actions/runs/14996740169/job/42132740449#step:7:36190):
```
[INFO] Mimir session closed (RETRIEVED=0 CACHED=2972)
```
So at least we know the numbers. We want `RETRIEVED=2972 CACHED=0`

cstamas committed:
Drop mixed bag caches

The "maven" cache caches local repository
that contains all sort of stuff, that in
reality you do NOT want cached.

joakime commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -71,6 +71,20 @@ jobs:
+        uses: actions/cache@v4
+        with:
+          path: ~/.mimir/local
+          key: mimir-codeql-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            mimir-codeql-
+            mimir-
```
the key being based on the poms is a good idea.  makes caching with multiple PRs and branches no big deal.

cstamas commented:
Ok, we had CodeQL in 30 minutes, pushing now next change and let's see.
Mimir reported `[INFO] Mimir session closed (RETRIEVED=0 CACHED=2979)`

cstamas committed:
Employ MBC and up wrapper plugin version

cstamas committed:
Trigger build again

cstamas committed:
MBC kills CodeQL

joakime commented:
Jetty 12.0.x has maven cache already enabled.

cstamas commented:
> Jetty 12.0.x has maven cache already enabled.

Yup, but it is not used nor cannot be used in CodeQL job it seems :disappointed:

cstamas commented:
So conclusion: not helped much. Local repo cache was already present, Mimir now "properly caches" (built and installed artifacts are NOT cached), and RRF cut down from build time but is not visible (due caching, with empty local repo and caches it IS visible very much).

Q: if CodeQL requires "just a quick build", why not use something like [Takari Smart Builder](https://github.com/takari/takari-smart-builder) only in this build? It does cut down build times just fine.

joakime commented:
> Q: if CodeQL requires "just a quick build", why not use something like [Takari Smart Builder](https://github.com/takari/takari-smart-builder) only in this build? It does cut down build times just fine.

_"Just a quick build"_ isn't a factor IMO.  
I was just trying to understand how to leverage this proposed feature in a Github Action, and used CodeQL as an example.
Which you have demonstrated rather well, thank you for that.

For now, we'll review the changes here and see if we can leverage this properly.

olamy commented:
We have very limited effect of this in our Jenkins build as we are using Nexus as mirror.
But for local build rrf is good idea .

cstamas commented:
Yup, my first build was dead slow as all these reposes were queried. Moreover they served things present in Central. Afterwards, with caches populated (Mimir or Nexus) it was not so apparent.

olamy reviewed (approved):
happy to merge it the RRF setup is great
